criu/proc_parse.c

```diff
@@ -206,6 +208,20 @@ static void parse_vma_vmflags(char *buf, struct vma_area *vma_area)
 
 	if (vma_area->e->madv)
 		vma_area->e->has_madv = true;
+
+	/*
+	 * We set MAP_PRIVATE flag on vma_area->e->flags right after parsing
+	 * a first line of VMA entry in /proc/<pid>/smaps file:
+	 * 7fa84fa70000-7fa84fa95000 rw-p 00000000 00:00 0
+	 * but it's too early and we can't distinguish between MAP_DROPPABLE
+	 * and MAP_PRIVATE mappings yet, as they both private mappings in nature
+	 * and at this point we haven't yet read "VmFlags:" line in smaps.
+	 *
+	 * Let's detect this situation and drop MAP_PRIVATE flag while keep
+	 * MAP_DROPPABLE, otherwise restorer's restore_mapping() helper will fail.
+	 */
+	if ((vma_area->e->flags & MAP_PRIVATE) && (vma_area->e->flags & MAP_DROPPABLE))
+		vma_area->e->flags &= ~MAP_PRIVATE;
```

Comment:
At first, I'm quite concerned that this makes MAP_PRIVATE not set in `vma_area->e->flags` but `vma_area->e->status` might have VMA_ANON_PRIVATE as this is set in `handle_vma` already. Looking for the check of MAP_PRIVATE in `vma->e->flags` in the code, I only see 2 places `can_dump_with_memfd_hugetlb` and `vma_entry_can_be_lazy`
- can_dump_with_memfd_hugetlb: MAP_DROPPABLE cannot reach this path as MAP_DROPPABLE cannot be hugetlb backed
- vma_entry_can_be_lazy: must return false on MAP_DROPPABLE as we cannot use userfaultfd on MAP_DROPPABLE,. As we drop MAP_PRIVATE here, this will correctly return false

So this change looks good to me.

Reply:
yeah, I was thinking about this too. It is kinda tricky. Thanks for double checking and ensuring that everything is correct! ;-)